File: backend/app/schemas/agent_team.py
"""Safe API/read schemas for the Phase 19A agent-team analysis console."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, model_validator

from app.schemas.actionability import ReviewActionabilityStatus
from app.schemas.trade_review_workspace import TradeReviewPortfolioPreviewRequest
from app.services.agent_team.llm_provider import (
    AGENT_TEAM_ROLES,
    LLM_PROVIDER_STATUSES,
    AgentTeamRole,
    LLMProviderStatus,
    find_prohibited_llm_phrases,
)
from app.services.agent_team.output_safety import validate_llm_provider_output
from app.services.privacy import FORBIDDEN_TRADE_REVIEW_WORKSPACE_KEYS, find_forbidden_keys


AgentTeamRunStatus = Literal["completed", "partially_completed", "failed"]


class AgentTeamAnalysisPreviewRequest(TradeReviewPortfolioPreviewRequest):
    """Request for a stateless mock agent-team analysis preview."""

    model_config = ConfigDict(from_attributes=True, extra="forbid")


class AgentTeamRoleOutputRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    role_name: AgentTeamRole
    display_name: str
    status: Literal["completed", "unavailable", "skipped"]
    provider_status: LLMProviderStatus
    content_markdown: str | None
    is_mock: bool
    unavailable_reason: str | None = None


class AgentTeamStageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    stage: str
    status: str
    role_name: AgentTeamRole | None = None
    display_name: str | None = None
    provider_status: LLMProviderStatus | None = None
    unavailable_reason: str | None = None


class AgentTeamProviderWarningRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    code: str
    message: str


class AgentTeamAnalysisConsoleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    run_reference: str
    workflow_version: str
    run_status: AgentTeamRunStatus
    generated_at: datetime
    review_flow_label: str
    review_actionability_status: ReviewActionabilityStatus
    broker_snapshot_freshness: dict[str, object]
    market_quote_freshness: dict[str, object]
    deterministic_evidence_summary: dict[str, object]
    role_outputs: tuple[AgentTeamRoleOutputRead, ...]
    final_synthesis: str | None
    provider_warnings: tuple[AgentTeamProviderWarningRead, ...]
    stages: tuple[AgentTeamStageRead, ...]
    safety_flags: tuple[str, ...]

    @model_validator(mode="after")
    def analysis_console_payload_must_be_safe(self) -> "AgentTeamAnalysisConsoleRead":
        validate_agent_team_console_payload(self.model_dump(mode="python"))
        return self


def validate_agent_team_console_payload(payload: object) -> None:
    forbidden = find_forbidden_keys(payload, forbidden_keys=FORBIDDEN_TRADE_REVIEW_WORKSPACE_KEYS)
    if forbidden:
        blocked = ", ".join(sorted(forbidden))
        raise ValueError(f"agent-team console payload contains forbidden private fields: {blocked}")
    # Reuse provider phrase validation for user-visible text while allowing safe strategy labels in structural fields.
    prohibited = find_prohibited_llm_phrases(payload)
    if prohibited:
        blocked = ", ".join(sorted(prohibited))
        raise ValueError(f"agent-team console payload contains prohibited phrase(s): {blocked}")
    validate_llm_provider_output(_provider_validated_subset(payload), label="agent-team console text")


def _provider_validated_subset(payload: object) -> object:
    """Return generated text fields for hardened provider-output validation."""

    if not isinstance(payload, dict):
        return payload
    return {
        "role_outputs": [
            {
                "role_name": output.get("role_name"),
                "content_markdown": output.get("content_markdown"),
                "unavailable_reason": output.get("unavailable_reason"),
            }
            for output in payload.get("role_outputs", ())
        ],
        "final_synthesis": payload.get("final_synthesis"),
        "provider_warnings": payload.get("provider_warnings", ()),
        "safety_flags": payload.get("safety_flags", ()),
    }


def supported_agent_team_roles() -> tuple[str, ...]:
    return AGENT_TEAM_ROLES


def supported_provider_statuses() -> tuple[str, ...]:
    return LLM_PROVIDER_STATUSES

File: frontend/src/types/agentTeam.ts
/**
 * TypeScript mirror of backend/app/schemas/agent_team.py.
 *
 * Field names, optionality, Decimal→string semantics, and Literal unions
 * follow the backend exactly. Per P19A-T2..T5 deferred polish notes, the
 * structural Records (broker_snapshot_freshness / market_quote_freshness /
 * deterministic_evidence_summary) stay loose-typed (`Record<string, unknown>`)
 * to mirror the backend `dict[str, object]` shape. Do not invent typed fields.
 */
import type { TradeReviewPortfolioPreviewRequest } from "./tradeReview";
import type { ReviewActionabilityStatus } from "./tradeReview";
import type { PortfolioContextSelectionMode } from "./tradeReview";
import type { PortfolioScopeMode } from "./accountDetails";

/* ── Enums (mirror llm_provider.py + agent_team.py) ─────────────────────── */

export type AgentTeamRole =
  | "fundamentals_analyst"
  | "news_analyst"
  | "technical_analyst"
  | "risk_management_agent"
  | "portfolio_manager_agent";

export const AGENT_TEAM_STAGE_ORDER: readonly AgentTeamRole[] = [
  "fundamentals_analyst",
  "news_analyst",
  "technical_analyst",
  "risk_management_agent",
  "portfolio_manager_agent",
] as const;

export type LLMProviderStatus =
  | "ok"
  | "skipped"
  | "failed"
  | "rate_limited"
  | "quota_exceeded"
  | "provider_timeout"
  | "provider_auth_error"
  | "provider_unavailable"
  | "invalid_response"
  | "safety_validation_failed";

export type AgentTeamRunStatus = "completed" | "partially_completed" | "failed";

export type AgentTeamRoleOutputStatus = "completed" | "unavailable" | "skipped";

/* ── Request — same shape as TradeReviewPortfolioPreviewRequest ──────────── */
/* AgentTeamAnalysisPreviewRequest extends it with `extra="forbid"`. We send
   nothing more than the parent already allows. */

export type AgentTeamAnalysisPreviewRequest = TradeReviewPortfolioPreviewRequest;

/* ── Response — AgentTeamAnalysisConsoleRead 1:1 ─────────────────────────── */

export interface AgentTeamRoleOutputRead {
  role_name: AgentTeamRole;
  /** Backend-owned user-facing label (P25A-T11). Render verbatim. */
  display_name: string;
  status: AgentTeamRoleOutputStatus;
  provider_status: LLMProviderStatus;
  content_markdown: string | null;
  is_mock: boolean;
  unavailable_reason?: string | null;
}

export interface AgentTeamStageRead {
  stage: string;
  status: string;
  role_name?: AgentTeamRole | null;
  /** Backend-owned user-facing label (P25A-T11); null for non-persona stages. */
  display_name?: string | null;
  provider_status?: LLMProviderStatus | null;
  unavailable_reason?: string | null;
}

export interface AgentTeamProviderWarningRead {
  code: string;
  message: string;
}

export interface AgentTeamScopeSummaryRead {
  scope_present: boolean;
  portfolio_scope_mode: PortfolioScopeMode;
  portfolio_context_selection_mode: PortfolioContextSelectionMode | null;
  selected_context_present: boolean;
  included_account_count: number;
  excluded_account_count: number;
  review_account_present: boolean;
  account_level_feasibility_evaluated: boolean;
  scope_caveat_codes: string[];
}

export interface AgentTeamAnalysisConsoleRead {
  run_reference: string;
  workflow_version: string;
  run_status: AgentTeamRunStatus;
  generated_at: string;
  review_flow_label: string;
  review_actionability_status: ReviewActionabilityStatus;
  /** Loose-typed per backend `dict[str, object]`. Render verbatim as a
   *  key/value list; do not infer typed fields. */
  broker_snapshot_freshness: Record<string, unknown>;
  /** Loose-typed per backend `dict[str, object]`. Distinct scope from broker. */
  market_quote_freshness: Record<string, unknown>;
  /** Loose-typed per backend `dict[str, object]` — counts/categories only;
   *  the backend never includes values, holdings, or identifiers here. */
  deterministic_evidence_summary: Record<string, unknown>;
  /** Lossy reviewed scope summary only. It intentionally carries no account
   *  labels, refs, balances, holdings, quantities, provider IDs, or payloads. */
  scope_summary: AgentTeamScopeSummaryRead;
  role_outputs: AgentTeamRoleOutputRead[];
  final_synthesis: string | null;
  provider_warnings: AgentTeamProviderWarningRead[];
  stages: AgentTeamStageRead[];
  safety_flags: string[];
}
